sympy/combinatorics/free_group.py

```diff
@@ -0,0 +1,916 @@
+    is_FreeGroup = True
+    is_PermutationGroup = False
+
+    def __new__(cls, *args, **kwargs):
+        """
+        Since any number of arguments can be passed in the form of string,
+        hence `rank` is "not" used as argument.
+        """
+
+        obj = Basic.__new__(cls, *args, **kwargs)
+
+        if isinstance(args[0], int) and args[0] >= 0:
+
+            # (1) form of the API used here
+            if len(args) == 1:
+                obj._str = "f"
+
+            # (2) form of the API used here
+            elif len(args) == 2:
+                if not isinstance(args[1], str):
+                    raise ValueError("second argument should be of type `str`, "
+                            "not of type %s" % type(args[1]))
+                obj._str = args[1]
+
+            else:
+                raise ValueError("Invalid arguments")
+
+            obj._rank = args[0]
+
+        # (3) form of API, where all the generators of `FreeGroup`
+        # given as strings.
+        elif all([isinstance(i, str) for i in args]):
+            pass
+
+        # (4) form of API, not sure how this should be implemented
+        # right now.
+        elif args[0] is S.Infinity:
+            pass
+
+        return obj
+
+    @property
+    def generators(self):
+        """Returns the generators of the FreeGroup
+
+        Examples
+        ========
+
+        >>> f = FreeGroup( 3, "swapnil" )
+        >>> f.generators
+        [swapnil0, swapnil1, swapnil2]
+
+        """
+        return [FreeGroupElm(self, [(i, 1)], self.as_str) for i in
+                range(self.rank)]
+
+    def __getitem__(self, i):
+        return self.generators[i]
+
+    def __contains__(self, i):
+        """
+        Return True if `i` is contained in FreeGroup.
+
+        Examples
+        ========
+
+        >>> from sympy import FreeGroup
+        >>> f = FreeGroup( 4, "swap" )
+        >>> g = FreeGroup( 4, "swapnil" )
+
+        >>> f[0]**2*f[3] in f
+        True
+        >>> f[0]**2*f[3] in g
+        False
+
+        """
+        if not isinstance(i, FreeGroupElm):
+            raise TypeError("FreeGroup contains elements of type "
+                    "`FreeGroupElm`")
+        return self.contains(i)
+
+    def __len__(self):
+        return self.rank
+
+    def __str__(self):
+        if self.rank > 30:
+            str_form = "<free group with %s generators>" % self.rank
+        else:
+            str_form = "<free group on the generators "
+            gens = self.generators
+            str_form += str(gens) + ">"
+        return str_form
+
+    __repr__ = __str__
+
+    def __eq__(self, other):
+        """No ``FreeGroup`` is equal to any "other" ``FreeGroup``.
+
+        Examples
+        ========
+
+        >>> from sympy import FreeGroup
+        >>> f = FreeGroup( 4, "swapnil" )
+        >>> g = FreeGroup( 4, "swapnil" )
+        >>> f == g
+        False
+        >>> f == f
+        True
+
+        """
+        return self is other
```

Comment:
> it is probable that Python would provide this as a default

But for the above doctest i have added. It is such that even if the two instances `f` and `g` have the same arguments (i.e "4" and "swapnil") but still `f == g` has to be returned as being not equal, different from the usual way Python defines `__eq__`(which returns `True` when the arguments are same for two instances of same class).